pyo3-build-config/src/impl_.rs

```diff
@@ -0,0 +1,882 @@
+        }
+
+        if self.is_pypy() {
+            println!("cargo:rustc-cfg=PyPy");
+            if self.abi3 {
+                warn!(
+                    "PyPy does not yet support abi3 so the build artifacts will be version-specific. \
+                    See https://foss.heptapod.net/pypy/pypy/-/issues/3397 for more information."
+                )
+            }
+        };
+
+        for flag in &self.build_flags.0 {
+            println!("cargo:rustc-cfg=py_sys_config=\"{}\"", flag)
+        }
+    }
+
+    pub fn is_pypy(&self) -> bool {
+        self.implementation == PythonImplementation::PyPy
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct PythonVersion {
+    pub major: u8,
+    pub minor: u8,
+}
+
+impl PythonVersion {
+    const PY37: Self = PythonVersion { major: 3, minor: 7 };
+}
+
+impl Display for PythonVersion {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}.{}", self.major, self.minor)
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum PythonImplementation {
+    CPython,
+    PyPy,
+}
+
+impl FromStr for PythonImplementation {
+    type Err = Box<dyn std::error::Error>;
+    fn from_str(s: &str) -> Result<Self> {
+        match s {
+            "CPython" => Ok(PythonImplementation::CPython),
+            "PyPy" => Ok(PythonImplementation::PyPy),
+            _ => bail!("Invalid interpreter: {}", s),
+        }
+    }
+}
+
+fn is_abi3() -> bool {
+    cargo_env_var("CARGO_FEATURE_ABI3").is_some()
+}
+
+trait GetPrimitive {
+    fn get_bool(&self, key: &str) -> Result<bool>;
+    fn get_numeric<T: FromStr>(&self, key: &str) -> Result<T>;
+}
+
+impl GetPrimitive for HashMap<String, String> {
+    fn get_bool(&self, key: &str) -> Result<bool> {
+        match self
+            .get(key)
+            .map(|x| x.as_str())
+            .ok_or(format!("{} is not defined", key))?
+        {
+            "1" | "true" | "True" => Ok(true),
+            "0" | "false" | "False" => Ok(false),
+            _ => bail!("{} must be a bool (1/true/True or 0/false/False", key),
+        }
+    }
+
+    fn get_numeric<T: FromStr>(&self, key: &str) -> Result<T> {
+        self.get(key)
+            .ok_or(format!("{} is not defined", key))?
+            .parse::<T>()
+            .map_err(|_| format!("Could not parse value of {}", key).into())
+    }
+}
+
+struct CrossCompileConfig {
+    lib_dir: PathBuf,
+    version: Option<String>,
+    os: String,
+    arch: String,
+}
+
+fn cross_compiling() -> Result<Option<CrossCompileConfig>> {
+    let cross = env_var("PYO3_CROSS");
+    let cross_lib_dir = env_var("PYO3_CROSS_LIB_DIR");
+    let cross_python_version = env_var("PYO3_CROSS_PYTHON_VERSION");
+
+    let target_arch = cargo_env_var("CARGO_CFG_TARGET_ARCH");
+    let target_vendor = cargo_env_var("CARGO_CFG_TARGET_VENDOR");
+    let target_os = cargo_env_var("CARGO_CFG_TARGET_OS");
```

Comment:
Unfortunately maturin can't not use Cargo env var since it's building other crates not itself. See https://github.com/PyO3/maturin/blob/8986455758125328744c8ab9ba3a2371d489860a/src/cross_compile.rs#L12-L13

Reply:
Mmm good point, I guess we'll need to refactor some of these functions to take target / arch / vendor as arguments? What form would be convenient for maturin?

Reply:
> I guess we'll need to refactor some of these functions to take target / arch / vendor as arguments?

That should work for maturin.